src/utils/logging.js

```diff
@@ -21,6 +23,43 @@ function withScope<R>(callback: Scope => R): R {
   return ((ret: $FlowFixMe): R);
 }
 
+/**
+ * Get server-version tags at various levels of granularity.
+ */
+const getServerVersionTags = () => {
+  const result = {};
+
+  const zulipVersion = tryGetActiveAccount(store.getState())?.zulipVersion;
+
+  // If there's no active account, we don't know the server version.
+  if (!zulipVersion) {
+    return result;
+  }
+
+  const raw = zulipVersion.raw();
+  result.rawServerVersion = raw;
+
+  const OMITTED = 'x';
+  const UNKNOWN = '?';
+
+  const elements = zulipVersion.elements();
+  const major = elements.major ?? UNKNOWN;
+  const minor = elements.minor ?? UNKNOWN;
+  const patch = elements.patch ?? UNKNOWN;
+
+  // Effective with 3.0, we changed our numbering conventions; 3.x and
+  // 4.x are each the same level of granularity as 2.1.x or 2.0.x.
+  if (zulipVersion.isAtLeast('3.0')) {
+    result.coarseServerVersion = [major, OMITTED].join('.');
+    result.fineServerVersion = [major, minor].join('.');
+  } else {
+    result.coarseServerVersion = [major, minor, OMITTED].join('.');
+    result.fineServerVersion = [major, minor, patch].join('.');
+  }
+
+  return result;
```

Comment:
Let's put this in a form where it ends with something like
```js
return { rawServerVersion: raw, coarseServerVersion, fineServerVersion };
```
That way when you're looking down at the `setTags` call, and trying to see what are all the tags our code sets here, when you look up at this function you only have to glance at the `return` statements to see the inventory of them. (And then if there's one you're interested in knowing more about, you can follow up to where that value gets computed.)

Reply:
What do you think about something like this:

```javascript
  // If there's no active account, we don't know the server version.
  if (zulipVersion) {
    // [...]
    return { rawServerVersion: raw, coarseServerVersion, fineServerVersion };
  } else {
    return {};
  }
```

That way, we can maintain that
- all values in the returned object are strings (none are undefined)
- it's quite easy for Flow to know the return type; either the returned object will have those three properties, or it will be empty.

Reply:
Yeah, those are good things to maintain. Looks good except let's make the quick `return {}` case an early return, to avoid letting that edge-case tail wag the dog of how the main case is formatted and read. Also keeps the edge-case return next to the condition that leads to it.